pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,202 @@
+	}
+	kasp = keepalive.ServerParameters{
+		Time:    5 * time.Second, // Ping the client if it is idle for 5 seconds to ensure the connection is still active
+		Timeout: 1 * time.Second, // Wait 1 second for the ping ack before assuming the connection is dead
+	}
+)
+
+// StreamPolicyState is a bidirectional streaming RPC to subscribe to DNS policies
+// SDP calls this method to subscribe to DNS policies
+// For each stream, we start a goroutine to receive the DNS policies ACKs
+// The flow of the method is as follows:
+// 1. Add the stream to the map( called by the client i.e SDP)
+// 2. Start a goroutine to receive the DNS policies ACKs for that particular client.
+// 3. Send the current state of the DNS rules to the client (We store the current state fo DNS rules during the endpoint regeneration see UpdatePolicyRulesLocked)
+// 4. Wait for the context to be done
+func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
+	return nil
+}
+
+// NewServer creates a new FQDNDataServer which is used to handle the Standalone DNS Proxy grpc service
+func NewServer(endpointManager endpointmanager.EndpointManager, updateOnDNSMsg updateOnDNSMsgFunc, logger *slog.Logger) *FQDNDataServer {
+	ctx := context.Background()
+
+	s := &FQDNDataServer{
+		endpointManager:     endpointManager,
+		updateOnDNSMsg:      updateOnDNSMsg,
+		ctx:                 ctx,
+		currentIdentityToIp: make(map[identity.NumericIdentity][]net.IP),
+		log:                 logger.With(logfields.LogSubsys, "fqdn/server"),
+	}
+
+	go func() {
+		<-s.ctx.Done()
+		s.log.Info("FQDN service context done, cleaning up resources")
+		if s.ctx.Err() != nil {
+			s.log.Error("FQDN service context error: ", logfields.Error, s.ctx.Err())
+		}
+		s.closeServerMu.Lock()
+		closeFunc := s.closeServer
+		s.closeServerMu.Unlock()
+		if closeFunc != nil {
+			closeFunc()
+		}
+	}()
+
+	return s
+}
+
+// OnIPIdentityCacheChange is a method to receive the IP identity cache change events
+func (s *FQDNDataServer) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr types.PrefixCluster, oldHostIP, newHostIP net.IP, oldID *ipcache.Identity, newID ipcache.Identity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata, endpointFlags uint8) {
+	s.identityToIpMutex.Lock()
+	defer s.identityToIpMutex.Unlock()
+
+	if cidr.ClusterID() == 0 {
+
+		ipNet := cidr.AsIPNet()
+		ones, _ := ipNet.Mask.Size()
+
+		if (ipNet.IP.To4() != nil && ones != 32) || (ipNet.IP.To4() == nil && ones != 128) {
+			s.log.Info("CIDR mask not supported", logfields.CIDR, cidr.String())
+		} else {
+			switch modType {
+			case ipcache.Upsert:
+				if oldID != nil {
+					// Remove from the old identity
+					if ips, ok := s.currentIdentityToIp[oldID.ID]; ok {
+						for i, ip := range ips {
+							if ip.Equal(ipNet.IP) {
+								s.currentIdentityToIp[oldID.ID] = append(ips[:i], ips[i+1:]...)
+								break
+							}
+						}
```

Comment:
Please use one of the modern `slices.Delete` helper here. Also you probably want to delete `s.currentIdentityToIp[oldID.ID]` if the slice is empty to avoid a memory leak.

Same for the `ipcache.Delete` case below. You might even want to introduce a helper function here.